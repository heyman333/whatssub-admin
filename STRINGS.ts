const STRINGS = {
  en: { // English
    BUTTON: 'Button',
    SIGNUP: 'Sign Up',
    LOGIN: 'Login',
    LOGOUT: 'Logout',
    GOOGLE_LOGIN: 'Login with Google',
    FACEBOOK_LOGIN: 'Login with Facebook',
    EMAIL: 'Email',
    EMAIL_HINT: 'Write email address.',
    EMAIL_ERROR: 'Wrong email address',
    PASSWORD: 'Password',
    PASSWORD_HINT: 'Write password.',
    COMPLETE: 'Done',
    GOTO_Temp: 'No Page',
    BACK: 'Back',
    NAVIGATE: 'Navigate',
    CHANGE_THEME: 'Change theme',
    SIGN_IN_SCREEN: 'Sign in screen',
    MAIN_SCREEN: 'Main screen',
    UPDATE_BUTTON: 'Update',
    DELETE_BUTTON: 'Delete',
    SERVICE_LIST: 'Service List',
    SERVICE_ADD_BUTTON: 'Add Service',
    PRODUCT_LIST: 'Product List',
    PRODUCT_ADD_BUTTON: 'Add Product',
    SAVE: 'Save',
    CANCEL: 'Cancel',
    ICON: 'Icon',
    SERVICE_NAME: 'Service Name',
    SERVICE_NAME_KOR: 'Service Name(KOR)',
    WEB_PAGE: 'Web Page',
    DESCRIPTION: 'Description',
    PNG_GIF_ONLY: 'you can upload png and gif files only',
    INPUT_REQUIRED: 'is required',
    INPUT_MAX_VALIDATION: 'you exceeded max length of',
  },
  ko: { // Korean
    BUTTON: '버튼',
    SIGNUP: '회원가입',
    LOGIN: '로그인',
    LOGOUT: '로그아웃',
    GOOGLE_LOGIN: '구글 로그인',
    FACEBOOK_LOGIN: '페이스북 로그인',
    EMAIL: '이메일',
    EMAIL_ERROR: '이메일이 잘못 입력되었습니다.',
    EMAIL_HINT: '이메일 주소를 입력해주세요.',
    PASSWORD: '비밀번호',
    PASSWORD_HINT: '비밀번호를 입력해주세요.',
    COMPLETE: '완료',
    GOTO_Temp: '없는 페이지',
    BACK: '뒤로',
    NAVIGATE: '이동하기',
    CHANGE_THEME: '테마 변경',
    SIGN_IN_SCREEN: '로그인 화면',
    MAIN_SCREEN: '메인 화면',
    UPDATE_BUTTON: '수정',
    DELETE_BUTTON: '삭제',
    SERVICE_LIST: '서비스 목록',
    SERVICE_ADD_BUTTON: '서비스 추가',
    PRODUCT_LIST: '상품 목록',
    PRODUCT_ADD_BUTTON: '상품 추가',
    SAVE: '저장',
    CANCEL: '취소',
    ICON: '아이콘',
    SERVICE_NAME: '서비스명',
    SERVICE_NAME_KOR: '서비스명(한글)',
    WEB_PAGE: '웹페이지',
    DESCRIPTION: '설명',
    PNG_GIF_ONLY: 'png파일과 gif파일만 업로드 가능합니다',
    INPUT_REQUIRED: '의 입력이 필요합니다',
    INPUT_MAX_VALIDATION: '해당 길이 이상으로 입력될 수 없습니다:',
  },
  // ja: { // Japanese
  // },
  // zh: { // Chinese
  // },
  // es: { // Spanish
  // },
  // fr: { // French
  // },
  // id: { // Indonesian
  // },
};

const DEFAULT_LANG = 'en';
let lang = DEFAULT_LANG;

if (navigator) {
  lang = navigator.language.substr(0, 2);
}

export const getString = (str: string) => {
  str = STRINGS[lang]
    ? STRINGS[lang][str]
    : STRINGS[DEFAULT_LANG][str];

  if (!str) {
    str = '...';
  }
  return str;
};

export default STRINGS;
