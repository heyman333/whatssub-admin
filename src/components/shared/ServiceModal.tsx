import React, { ComponentType, useCallback, forwardRef, useState } from 'react';
import styled from 'styled-components';
import { Formik, FormikProps, Field } from 'formik';
import * as Yup from 'yup';
import SimpleButton from './SimpleButton';
import { getString } from '../../../STRINGS';
import { IService } from '../../types';

export interface IServiceModalProviderState {
  show?: boolean;
  id?: string;
  name?: string;
  nameKr?: string;
  icon?: string;
  url?: string;
  description?: string;
};
export interface IServiceModalProviderPayload {
  show?: boolean;
  id?: string;
  name?: string;
  nameKr?: string;
  icon?: string;
  delState?: Array<string>;
  deleteMode?: boolean;
};

interface ISCWrapperProps {
  show: boolean;
}
interface IProps {
  onAddServiceClick?: () => void;
  onServiceClick?: (serviceId: string) => void;
  onUpdateServiceClick?: (serviceInfo: IServiceModalProviderState) => void;
  onDeleteServiceClick?: (serviceId: string) => void;
  onCloseServiceClick: () => void;
  serviceInfo: IServiceModalProviderState;
  onSubmitClick?: (service: IService) => void;
  onHandleImgChange?: () => void;
}
interface IHandleImageChange {
  file: Blob;
  setFieldValue?: any;
  field?: string;
  isPassed?: boolean;
}

const WhenNarrow = `
  @media only screen and (max-width: 760px),
  (min-device-width: 768px) and (max-device-width: 1024px)
`;
const inputCommonStyle = `
  width: 100%;
  border-radius: .3rem;
  padding: 0.5em;
  background: papayawhip;
  border: none;
  font-size: 1rem;
`;
const flexMiddle = `
  justify-content: center;
  align-items: center;
`;

const Wrapper = styled.div`
  background-color: rgba(0, 0, 0, 0.5);
  position: absolute;
  top: 0;
  left: 0;
  height: 100vh;
  width: 100vw;
  overflow: scroll;
  display: flex;
  opacity: ${({ show }: ISCWrapperProps) => (show ? 1 : 0)};
  flex-direction: column;
  ${flexMiddle}
  transition: visibility 0s, opacity 0.2s linear;
  z-index: 9999;
  visibility: ${({ show }: ISCWrapperProps) => (show ? 'visible' : 'hidden')};
`;
const Container = styled.div`
  background: ${(props) => props.theme.background};
  color: ${(props) => props.theme.fontColor};
  align-self: stretch;
  padding: 2rem;
  margin: 2rem;
  overflow: scroll;
  border-radius: 0.5rem;
  box-shadow: 2px 4px 0.5rem 0 #323232;
`;
const InputContentArea = styled.div`
  margin: 1rem 0;
`;
const RowAlignRight = styled.div`
  text-align: right;
`;
const ShortInputBox = styled.input`
  ${inputCommonStyle}
`;
const LongInputBox = styled.textarea`
  ${inputCommonStyle}
`;

const InputRowRoot = styled.div`
  display: flex;
  flex-direction: row;
  margin-bottom: 0.2rem;
  ${WhenNarrow} {
    display: block;
  }
`;
const InputRowKey = styled.div`
  flex: 1;
`;
const InputRowValue = styled.div`
  flex: 2;
`;
const PreviewArea = styled.div`
  display: flex;
  ${flexMiddle}
  margin: .5rem;
`;
const PreviewImgArea = styled.div`
  width: 5rem;
  height: 5rem;
  background-color: papayawhip;
  display: flex;
  flex-direction: column;
  ${flexMiddle}
`;
const PreviewImg = styled.img`
  width: 100%;
  height: 100%;
`;
const FileInput = styled.input`
  width: 100%;
`;
const FileInputArea = styled.div`
  display: flex;
  flex: 3;
  flex-direction: column;
  justify-content: center;
`;
const AlertArea = styled.div`
  width: 100%;
  padding: 0.2rem;
  font-size: 0.8rem;
  color: red;
`;

const ServiceModal = forwardRef<HTMLDivElement, IProps>((props, ref) => {
  const {
    onUpdateServiceClick, onCloseServiceClick, onSubmitClick, onHandleImgChange, serviceInfo,
  } = props;
  const { show, id, name, nameKr, icon, url, description } = serviceInfo;
  const [imagePreviewSrc, setImagePreviewSrc] = useState(null);
  let initialValuesObj = {
    icon: icon,
    uploadIcon: null,
    serviceName: name || '',
    serviceNameKor: nameKr || '',
    url: url || '',
    description: description || '',
  };
  const SUPPORTED_FORMATS = ['image/gif', 'image/png'];
  const maxInputLength = 100;
  const maxLenValidation =
    getString('INPUT_MAX_VALIDATION') + ' ' + maxInputLength;

  const onCloseEvt = useCallback(() => {
    setImagePreviewSrc(null);
    onCloseServiceClick();
  }, [show]);
  const AlertAreaDom = ({ errorMsg }) => {
    return !errorMsg ? null : <AlertArea>{errorMsg}</AlertArea>;
  };
  const handleImageChange = ({
    file,
    setFieldValue,
    field,
  }: IHandleImageChange) => {
    if (file) {
      setFieldValue(field, file);
      const isPassed = imageChangeValidation(file);
      if (!isPassed) return;
      let reader = new FileReader();
      reader.onloadend = () => {
        setImagePreviewSrc(reader.result);
      };
      reader.readAsDataURL(file);
    }
  };
  const imageChangeValidation = (file) => {
    let iconExt = icon && 'image/' + icon.substring(icon.lastIndexOf('.') + 1);
    let target = file || { type: iconExt };
    return target && SUPPORTED_FORMATS.includes(target.type);
  };

  const formOnSubmit = (values, { setSubmitting }) => {
    setTimeout(() => {
      console.log('submit! values', values);
      alert(JSON.stringify(values, null, 2));
      onSubmitClick({
        id,
        name: values.serviceName,
        nameKr: values.serviceNameKor,
        icon,
        url: values.url,
        description: values.description,
      });
    }, 400);
  };
  const ServiceContentsInputDom = ({
    handleSubmit,
    handleChange,
    handleBlur,
    setFieldValue,
    touched,
    values,
    errors,
  }: FormikProps<any>) => {
    const inputFileOnChange = (e) => {
      let file = e.target.files[0];
      let handleImageFn = handleImageChange;
      // let handleImageFn = onHandleImgChange || handleImageChange;
      handleImageFn({ file, setFieldValue, field: 'uploadIcon' });
    };
    const imgSrc = !imagePreviewSrc ? values.icon : imagePreviewSrc;
    return (
      <form onSubmit={handleSubmit}>
        <h3>{getString('SERVICE_ADD_BUTTON')}</h3>
        <InputContentArea>
          <InputRowRoot>
            <InputRowKey>{getString('ICON')}</InputRowKey>
            <InputRowValue>
              <InputRowRoot>
                <PreviewArea>
                  <PreviewImgArea data-testid={'imagePreview'}>
                    {imgSrc ? (
                      <PreviewImg src={imgSrc} data-testid={'imageArea'} />
                    ) : (
                      <div data-testid={'defaultPreview'}>
                        <div>png / gif</div>
                        <div>512 x 512</div>
                      </div>
                    )}
                  </PreviewImgArea>
                </PreviewArea>
                <FileInputArea>
                  <FileInput
                    data-testid={'icon'}
                    type="file"
                    name="uploadIcon"
                    onChange={inputFileOnChange}
                  />
                  <AlertAreaDom errorMsg={errors.uploadIcon} />
                </FileInputArea>
              </InputRowRoot>
            </InputRowValue>
          </InputRowRoot>
          <InputRowRoot>
            <InputRowKey>{getString('SERVICE_NAME')}</InputRowKey>
            <InputRowValue>
              <ShortInputBox
                data-testid={'serviceName'}
                type="text"
                name="serviceName"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.serviceName}
              />
              {touched.serviceName && (
                <AlertAreaDom errorMsg={errors.serviceName} />
              )}
            </InputRowValue>
          </InputRowRoot>
          <InputRowRoot>
            <InputRowKey>{getString('SERVICE_NAME_KOR')}</InputRowKey>
            <InputRowValue>
              <ShortInputBox
                data-testid={'serviceNameKor'}
                type="text"
                name="serviceNameKor"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.serviceNameKor}
              />
              {touched.serviceNameKor && (
                <AlertAreaDom errorMsg={errors.serviceNameKor} />
              )}
            </InputRowValue>
          </InputRowRoot>
          <InputRowRoot>
            <InputRowKey>{getString('WEB_PAGE')}</InputRowKey>
            <InputRowValue>
              <ShortInputBox
                data-testid={'url'}
                type="text"
                name="url"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.url}
              />
              {touched.url && <AlertAreaDom errorMsg={errors.url} />}
            </InputRowValue>
          </InputRowRoot>
          <InputRowRoot>
            <InputRowKey>{getString('DESCRIPTION')}</InputRowKey>
            <InputRowValue>
              <LongInputBox
                data-testid={'description'}
                name="description"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.description}
              />
              {touched.description && (
                <AlertAreaDom errorMsg={errors.description} />
              )}
            </InputRowValue>
          </InputRowRoot>
        </InputContentArea>
        <RowAlignRight>
          <SimpleButton
            data-testid={`serviceModalCloseBtn`}
            onClick={onCloseEvt}
          >
            {getString('CANCEL')}
          </SimpleButton>
          <SimpleButton data-testid={'serviceSubmit'} type="submit">
            {getString('SAVE')}
          </SimpleButton>
        </RowAlignRight>
      </form>
    );
  };
  return (
    <Wrapper ref={ref} show={show}>
      <Container>
        <Formik
          onSubmit={formOnSubmit}
          initialValues={initialValuesObj}
          enableReinitialize={true}
          validationSchema={Yup.object().shape({
            uploadIcon: Yup.mixed().test(
              'fileFormat',
              getString('PNG_GIF_ONLY'),
              imageChangeValidation
            ),
            serviceName: Yup.string()
              .max(maxInputLength, maxLenValidation)
              .required(
                getString('SERVICE_NAME') + ' ' + getString('INPUT_REQUIRED')
              ),
            serviceNameKor: Yup.string()
              .max(maxInputLength, maxLenValidation)
              .required(
                getString('SERVICE_NAME_KOR') + ' ' + getString('INPUT_REQUIRED')
              ),
            url: Yup.string()
              .max(maxInputLength, maxLenValidation)
              .required(getString('WEB_PAGE') + ' ' + getString('INPUT_REQUIRED')),
            description: Yup.string()
              .max(maxInputLength, maxLenValidation)
              .required(
                getString('DESCRIPTION') + ' ' + getString('INPUT_REQUIRED')
              ),
          })}
          component={ServiceContentsInputDom}
        />
      </Container>
    </Wrapper>
  );
});

export default ServiceModal;
