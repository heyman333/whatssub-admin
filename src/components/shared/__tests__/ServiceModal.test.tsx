import React from 'react';
import renderer, { ReactTestRendererJSON } from 'react-test-renderer';

import ServiceModal from '../ServiceModal';
import { render, fireEvent, RenderResult, waitForElement, act, waitForDomChange } from '@testing-library/react';
import { getString } from '../../../../STRINGS';

let component: React.ReactElement;
let testingLib: any;
const serviceMock = {
  id: '87355',
  name: 'Davis, Howell and Lindgren',
  nameKr: 'Fahey - Bradtke',
  icon: 'https://s3.amazonaws.com/uifaces/faces/twitter/primozcigler/128.jpg',
  url: 'http://whats.sub',
  description: 'whatssub',
};
let initProps = {
  show: false,
  icon: null,
  uploadIcon: null,
  serviceName: '',
  serviceNameKor: '',
  url: '',
  description: '',
};
const SUPPORTED_FORMATS = ['image/gif', 'image/png'];
const imageChangeValidation = (file) =>
  file && SUPPORTED_FORMATS.includes(file.type);

let props = {
  onAddServiceClick: () => jest.fn(),
  onServiceClick: jest.fn(),
  onUpdateServiceClick: jest.fn(),
  onDeleteServiceClick: jest.fn(),
  onCloseServiceClick: jest.fn(),
  onSubmitClick: jest.fn((arg) => arg),
  onHandleImgChange: jest.fn(),
  serviceInfo: {
    show: initProps.show,
    id: '',
    name: initProps.serviceName,
    nameKr: initProps.serviceNameKor,
    icon: initProps.icon,
  },
};

describe('[ServiceModal] render', () => {
  it('renders without crashing', () => {
    const rendered: ReactTestRendererJSON = renderer
      .create(<ServiceModal {...props} />)
      .toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });
});

describe('[ServiceModal] interactions', () => {
  const component = <ServiceModal {...props} />;
  let renderResult: RenderResult;

  beforeAll(() => {
    renderResult = render(component);
  });
  it('modal closed inline method triggered', () => {
    const modalCloseButton = renderResult.getByTestId('serviceModalCloseBtn');
    fireEvent.click(modalCloseButton);
    expect(props.onCloseServiceClick).toHaveBeenCalled();
  });
  it('file type check when user upload icon image file', async () => {
    let iconInput = renderResult.getByTestId('icon');
    const invalidFile = new File(['invalidfile'], 'invalidfile.jpg', {
      type: 'image/jpg',
    });
    fireEvent.change(iconInput, { target: { files: [invalidFile] } });
    const defaultPreview = await waitForElement(() => renderResult.getByTestId('defaultPreview'));
    expect(defaultPreview).toBeTruthy();

    const validFile = new File(['validfile'], 'validfile.png', {
      type: 'image/png',
    });
    fireEvent.change(iconInput, { target: { files: [validFile] } });
    const imageAreaAfterChange = await waitForElement(() => renderResult.getByTestId('imageArea'));
    expect(imageAreaAfterChange).toBeTruthy();
  });
  it('existed data is rendered well on modal', () => {
    let newProps = {
      ...props,
      serviceInfo: serviceMock,
    };
    renderResult.rerender(<ServiceModal {...newProps} />);

    const serviceNameInput = renderResult.getByTestId('serviceName') as HTMLInputElement;
    const serviceNameKorInput = renderResult.getByTestId('serviceNameKor') as HTMLInputElement;
    const webPageInput = renderResult.getByTestId('url') as HTMLInputElement;
    const descriptionInput = renderResult.getByTestId('description') as HTMLInputElement;

    expect(serviceNameInput.value).toBe(serviceMock.name);
    expect(serviceNameKorInput.value).toBe(serviceMock.nameKr);
    expect(webPageInput.value).toBe(serviceMock.url);
    expect(descriptionInput.value).toBe(serviceMock.description);
  });
  it('onSubmit inline method triggered', async () => {
    // TODO: get to be triggered
    const modalSaveButton = renderResult.getByTestId('serviceSubmit');
    fireEvent.submit(modalSaveButton);
    setTimeout(() => {
      expect(props.onSubmitClick).toHaveBeenCalled();
    }, 410);
  });
});
