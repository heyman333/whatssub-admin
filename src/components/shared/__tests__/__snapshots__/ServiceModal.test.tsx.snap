// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ServiceModal] render renders without crashing 1`] = `
<div
  className="sc-bwzfXH jyhetQ"
>
  <div
    className="sc-htpNat hWibBV"
  >
    <form
      onSubmit={[Function]}
    >
      <h3>
        Add Service
      </h3>
      <div
        className="sc-bxivhb ElINa"
      >
        <div
          className="sc-gzVnrw eGGpmP"
        >
          <div
            className="sc-htoDjs cIVlcQ"
          >
            Icon
          </div>
          <div
            className="sc-dnqmqq eoabIh"
          >
            <div
              className="sc-gzVnrw eGGpmP"
            >
              <div
                className="sc-iwsKbI cKuSGv"
              >
                <div
                  className="sc-gZMcBi fsWJbt"
                  data-testid="imagePreview"
                >
                  <div
                    data-testid="defaultPreview"
                  >
                    <div>
                      png / gif
                    </div>
                    <div>
                      512 x 512
                    </div>
                  </div>
                </div>
              </div>
              <div
                className="sc-jTzLTM fxOyQp"
              >
                <input
                  className="sc-VigVT kxLHnZ"
                  data-testid="icon"
                  name="uploadIcon"
                  onChange={[Function]}
                  type="file"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          className="sc-gzVnrw eGGpmP"
        >
          <div
            className="sc-htoDjs cIVlcQ"
          >
            Service Name
          </div>
          <div
            className="sc-dnqmqq eoabIh"
          >
            <input
              className="sc-EHOje fAHoIl"
              data-testid="serviceName"
              name="serviceName"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="sc-gzVnrw eGGpmP"
        >
          <div
            className="sc-htoDjs cIVlcQ"
          >
            Service Name(KOR)
          </div>
          <div
            className="sc-dnqmqq eoabIh"
          >
            <input
              className="sc-EHOje fAHoIl"
              data-testid="serviceNameKor"
              name="serviceNameKor"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="sc-gzVnrw eGGpmP"
        >
          <div
            className="sc-htoDjs cIVlcQ"
          >
            Web Page
          </div>
          <div
            className="sc-dnqmqq eoabIh"
          >
            <input
              className="sc-EHOje fAHoIl"
              data-testid="url"
              name="url"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="sc-gzVnrw eGGpmP"
        >
          <div
            className="sc-htoDjs cIVlcQ"
          >
            Description
          </div>
          <div
            className="sc-dnqmqq eoabIh"
          >
            <textarea
              className="sc-bZQynM euFUXE"
              data-testid="description"
              name="description"
              onBlur={[Function]}
              onChange={[Function]}
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="sc-ifAKCX gENqNG"
      >
        <button
          className="sc-bdVaJa kLNRSS"
          data-testid="serviceModalCloseBtn"
          onClick={[Function]}
        >
          Cancel
        </button>
        <button
          className="sc-bdVaJa kLNRSS"
          data-testid="serviceSubmit"
          type="submit"
        >
          Save
        </button>
      </div>
    </form>
  </div>
</div>
`;
